[INTERNAL] Add eslint to mocha test suite

IanVS opened the pull request:
Addresses https://github.com/artificialio/sane/issues/108

Standard `npm test` will now run eslint in `/bin`, `/lib`, and `/tests`, as well as unit and acceptance tests

Run `npm test -- lint` to run eslint alone
Run `npm test -- skip-lint` to only run unit and acceptance tests

Globegitter commented on .eslintrc:
```diff
@@ -4,20 +4,30 @@
     "strict": 0,
     "no-underscore-dangle": 0,
     "no-unused-vars": 0,
-    "curly": 0,
```
I am not a fan of inline ifs, so let's put curly in at least as a warning.

Globegitter commented on .eslintrc:
```diff
@@ -4,20 +4,30 @@
     "strict": 0,
     "no-underscore-dangle": 0,
     "no-unused-vars": 0,
```
Let's have this as an error.

Globegitter commented on .eslintrc:
```diff
@@ -4,20 +4,30 @@
     "no-unused-vars": 0,
-    "curly": 0,
     "no-multi-spaces": 0,
     "key-spacing": 0,
```
I haven't been consistent myself about that, but I do like the `"key-spacing": [2, { "align": "colon" }]` convention. Makes it imo easier to see what the values of an object are.

Globegitter commented on .eslintrc:
```diff
@@ -4,20 +4,30 @@
-    "curly": 0,
     "no-multi-spaces": 0,
     "key-spacing": 0,
     "no-return-assign": 0,
```
Let's have this on as well. Don't really see any point of assignments in return statements.

Globegitter commented:
Yep, this is really good to have! @kriswill @mgenev anything that you are totally against or think is missing? Now is your chance to chime in ;)

Globegitter commented:
@IanVS I think it was a good choice to set the `-- skip-lint` as default for now. Once we fixed the rules we can merge this in and have another PR to make the tests pass and remove the ` -- skip-lint` all-together. Having the `-- lint` is really nice though to just quickly check your style :)

Also would you be interested in writing your first own npm module similar to https://github.com/Muscula/mocha-jshint ? I think that would be pretty useful. We can put it under https://github.com/sane (if you want) and I could guide you through it :)

Globegitter commented:
Also is this rule http://eslint.org/docs/rules/eol-last.html on as default? Same for http://eslint.org/docs/rules/no-trailing-spaces.html?

I think we should also have `no-multiple-empty-lines`. I think usually one does not need more than 1 space, but am open to have `no-multiple-empty-lines: [1, {max: 2}]` if anyone has a strong preference.

IanVS commented:
@Globegitter Yes I was already thinking of creating a module like mocha-jshint, since there doesn't seem to be one for ESLint. Should be very simple to do. 

Regarding rule defaults, the main listing of rules (http://eslint.org/docs/rules/) will indicate if a rule is off by default. That's also why removing the rule for curly will enable it. 

As for multiple empty lines, the only time I like to have two lines is between the variable declarations at the top of the file and the functions blocks. Well, and maybe between function blocks too.  I like the max: 2 idea. That's just my personal style though.

Globegitter commented:
Awesome, yep then I will wait until you extracted it out into a npm module. And yeah now seeing the off by default note.

Ok then let's go for the `max: 2` rule. Als http://eslint.org/docs/rules/max-depth.html and http://eslint.org/docs/rules/max-len.html (though I think more than 80 nowadays) could be interesting.

Oh wow, there is even http://eslint.org/docs/rules/max-nested-callbacks.html. Anyway these seem to get fairly complex and I think what we have so far (including the comments) is a really good start. We can still tweak once merged and we got started on fixing the code.

Globegitter commented on .eslintrc:
```diff
@@ -4,20 +4,30 @@
+      "named": "never"
+    }],
+    "space-infix-ops": 2,
+    "spaced-line-comment": [1, "always", {exceptions:["-","+"]}]
```
Let's put quotes around exceptions, just so it is a completely valid JSON

IanVS replied:
Good catch.  I yanked this straight from the docs (http://eslint.org/docs/rules/spaced-line-comment.html) so I also created https://github.com/eslint/eslint/pull/2019 to update the docs.

Globegitter commented:
Ok, might have to leave key-spacing off for now: https://github.com/eslint/eslint/issues/2018 (or if I did miss something and this is possible than we can turn it on)

Globegitter commented on .eslintrc:
```diff
@@ -4,20 +4,30 @@
-    "no-cond-assign": 0,
-    "quotes": "single"
+    "comma-dangle": [1, "never"],
+    "no-use-before-define": 1,
```
Lets put this as an error for now.

IanVS commented on .eslintrc:
```diff
@@ -4,20 +4,30 @@
-    "new-parens": 0,
-    "no-cond-assign": 0,
-    "quotes": "single"
+    "comma-dangle": [1, "never"],
```
@Globegitter How do you feel about `"comma-dangle": [2, "always"],`? 

This would ensure that all object literals and array literals end in a dangling comma (to prevent errors when adding items).  It is valid ECMAScript, and the only reason not to use them seems to be IE8, which we don't care about (http://eslint.org/docs/rules/comma-dangle.html).   The code is not written this way currently, so it would take a little work to make the test pass, but that's something I could tackle.  

```
var foo = {
    bar: "baz",
    qux: "quux",
};
```

Globegitter commented:
This is great and well needed, already catching some errors!

Globegitter commented:
Playing around with it more and I think we should fix the eslint errors before merging into master. So maybe once this is merged in the both of us can tackle this together on https://github.com/artificialio/sane/pull/114

Happy to then give you contributor rights, so we can both work on the same branch. This will also finally get me to remove https://github.com/artificialio/sane/blob/master/lib/execAbort.js

IanVS commented:
@Globegitter sure thing that sounds fine.  I'll finish up changes to this PR tonight.

IanVS commented:
OK I created mocha-eslint (https://github.com/BadgeLabs/mocha-eslint) and I think I have addressed the comments from @Globegitter.   If everyone is happy with it so far, I can squash the commits before you merge if you'd like since there are quite a few of them.  I have a feeling we'll be tweaking the ESLint rules a bit more, but at least this is a starting point.

Globegitter commented:
@IanVS Yep please squash the commits and otherwise that looks great, ready to be merged.

IanVS commented:
@Globegitter Squished and squashed.
